include/trieste/source.h

```diff
@@ -71,46 +71,97 @@ namespace trieste
 
     std::pair<size_t, size_t> linecol(size_t pos) const
     {
-      // Lines and columns are 0-indexed.
-      auto it = std::lower_bound(lines.begin(), lines.end(), pos);
+      // If we have no lines, contents is an empty string.
+      // Realistically this case will only happen for pos == 0.
+      if (lines.empty())
+      {
+        return {0, pos};
+      }
 
-      auto line = it - lines.begin();
-      auto col = pos;
+      // Find the first line that begins _after_ pos, and then backtrack one
+      // element if we need to. We can't write this directly because "last
+      // element for which condition was false" is not an stdlib primitive.
+      auto it = std::find_if_not(
+        lines.begin(), lines.end(), [&](std::pair<size_t, size_t> elem) {
+          return elem.first <= pos;
+        });
+      // If we're at the beginning already, or we couldn't find a line after
+      // pos, then our current line should be fine. If lines is constructed
+      // correctly, it is on either the only line or the last line (and pos is
+      // on that line or beyond all lines).
+      if (it == lines.end() || (it != lines.begin() && it->first > pos))
+      {
+        --it;
+      }
 
-      if (it != lines.begin())
-        col -= *(it - 1) + 1;
+      size_t line = it - lines.begin();
+      size_t col = pos - it->first;
 
       return {line, col};
     }
 
     std::pair<size_t, size_t> linepos(size_t line) const
     {
-      // Lines are 0-indexed.
-      if (line > lines.size())
-        return {std::string::npos, 0};
-
-      size_t start = 0;
-      auto end = contents.size();
-
-      if (line > 0)
-        start = lines[line - 1] + 1;
-
-      if (line < lines.size())
-        end = lines[line];
+      // Special case: for out of range lines, index them at the end of our
+      // string and give them length 0. This will cause minimally-ugly
+      // misbehavior if there's an out of range error. Also, this gracefully
+      // handles the technically out of range situation where we ask for line 0
+      // of an empty string, which is what linecol(0) will give its caller in
+      // that case.
+
+      // Change note: this used to return std::string::npos when line was out of
+      // range by more than 1, but callers weren't checking for it so that case
+      // just caused things like .subtr(max_long) in practice. Best return an
+      // empty line with max idx, which will cause blank outputs. Otherwise,
+      // actually assert(line <= lines.size()) to crash here and not half way
+      // down the calling function.
+      if (line >= lines.size())
+      {
+        return {contents.size(), 0};
+      }
 
-      return {start, end - start};
+      return lines[line]; // already in {start, size} format
     }
 
   private:
     void find_lines()
     {
-      // Find the lines.
-      auto pos = contents.find('\n');
+      using namespace std::string_view_literals;
```

Comment:
I'll push back on this one a bit, for a few reasons:
- This style is easier to reason about, in my opinion. It does only guarded lookaheads, and it's obvious from the in-code literals what patterns it's looking for.
  For instance, I'm pretty sure your proposed alternative will make an invalid memory access if given the string `"\n"`. It will find `'\n'` at index 0, then check `cursor[0-1] == '\r'`. Yes, I could fix that with more conditionals, but at that point my head hurts and I'm not sure if requiring that level of care is worth it.
- I checked in Godbolt, and the optimizer strips all strings/string views and lambda calls from the output, resulting in some kind of inscrutable state machine that only calls out to vector functions, or to throw runtime exceptions. Therefore, I don't buy that there is much performance impact from changing between these implementations (in either direction). See my setup: https://godbolt.org/z/EnaaPhfqM

Reply:
Fair enough. In that case, maybe a compromise (to avoid capturing local state and put all the increments in the same place):

```c++
auto is_eol = [](const std::string_view s) -> std::pair<bool, size_t> {
        if(s.size() > 1 && s.substr(0, 2) == "\r\n"){
            return {true, 2};
        }
        
        if(s.empty()){
            return {false, 0};
        }
       
        if(s[0] == '\n'){
            return {true, 1};
        }
        
        return {false, 1};
    };

    size_t cursor = 0;
    size_t line_start = 0;
    std::string_view contents_view(contents);
    while(cursor < contents_view.size()){
        auto [eol, inc] = is_eol(contents_view.substr(cursor));
        if(eol){
            lines.emplace_back(line_start, cursor);
            cursor += inc;
            line_start = cursor;
        }else{
            cursor += inc;
        }
    }

    if(cursor - line_start > 0){
        lines.push_back({line_start, cursor});
    }
```

Reply:
Moved to other PR.